### DISCUSSION TITLE
A little help with Long-Term Memory extension, please

### DISCUSSION THREAD

**Discussion by dastaten** [wawawario2/text-generation-webui#27] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/3282357/233825617-5934b78d-5228-41e4-b7c8-6a47776ba427.png)

I've got oobabooga running with Pygmalion, and it's working perfectly, but I'm having trouble understanding how to add the Long-Term Memory extension.

I haven't worked with the command prompt very much, and I'm confused about step 3. It says, “Within the textgen conda environment.” How do I get there? It mentions “from the linked instructions,” but I don't see the linked instructions.

And in what directory should I be when I perform step 4?

I apologize if these are dumb questions. I'm still pretty new to a lot of this. I really want to get this extension installed, so any help is greatly appreciated.

Thanks.

**Comment by Thellton** [comment, marked as answer]
to run those commands you need to run a cmd line interface in the C or similar drive:\what ever your precise directory is\oobabooga_windows\text-generation-webui. to do this on windows, navigate to this location using file explorer and then launch cmd.exe in this location. this is done by typing 'cmd' into the text input box that displays the current directory location and you will start with the cmd line operating in this directory. you can then copy the commands and then right click to paste them into the cmd line. I should note, right clicking in this context does not reveal a context sensitive right click menu but rather pastes whatever is in the copy and paste clipboard.

if anything doesn't make sense in the above, all you need do is ask for additional elaboration about whatever it is in my description that is confusing you as I must confess, I was in the same boat and it took me three hours of effort to figure out what the maintainer meant. if you're reading this wawawario2; definitely not your fault just us muddling along with our bare minimum of computing knowledge.

as for step 4, its something you need to do to a file called webui.py and the purpose of this instruction is telling oobabooga webui to always launch with this extension operating. if you created your installation of oobabooga with the one click installer, this is a file sitting in the same directory as the start_windows.bat file. I would recommend downloading notepad++ for this as you will be adding '--extensions long_term_memory' to a line of code in webui.py at line 146 which typically looks like the following.

> run_cmd("python server.py --chat --model-menu) # put your flags here!

**Reply by dastaten** [reply]
Thanks for your help. So, I finished all the steps, but when I run start_windows.bat, I get the following:

Loading the extension "long_term_memory"... Fail.
Traceback (most recent call last):
  File "E:\oobabooga_windows\text-generation-webui\modules\extensions.py", line 19, in load_extensions
    exec(f"import extensions.{name}.script")
  File "<string>", line 1, in <module>
  File "E:\oobabooga_windows\text-generation-webui\extensions\long_term_memory\script.py", line 14, in <module>
    from extensions.long_term_memory.core.memory_database import LtmDatabase
  File "E:\oobabooga_windows\text-generation-webui\extensions\long_term_memory\core\memory_database.py", line 8, in <module>
    from sentence_transformers import SentenceTransformer
ModuleNotFoundError: No module named 'sentence_transformers'

Everything installed with no issues, and the long_term_memory directory is in my extensions folder.

Thanks again for the help.

**Reply by Thellton** [reply]
glad to have been of help, unfortunately my helpfulness basically is nil with this traceback so I think you'll have to hope wawawario2 finds this :(

**Reply by PalaSerialX** [reply]
You need to pip install sentence_transformers, its missing that module. if you go to the Oobabooga directory in the command line and just pip install sentence_transformers it will install the required library. 

as side note: You can feed ChatGPT this type of errors and it will help guide you on what might be missing or the issue, in this case its pretty simple. Its just missing the sentance_transformers library.

**Reply by dastaten** [reply]
Thanks, I tried that, and I'm still getting the same error. Steps I took:

1. Navigated to E:\oobabooga_windows in file explorer
2. Typed 'cmd' in the window to open a command prompt
3. Ran 'pip install sentence_transformers'

After the install, there was no difference in the E:\oobabooga_windows directory, and I noticed the file paths during the install were all c:\users\dorian\appdata\local\programs\python\python310\lib\site-packages

Does that mean it's installing in the wrong location?

Thanks for your help.

**Reply by PalaSerialX** [reply]
that's correct you want to install it into your oobabooga folder. So you want to navigate command  to cd to your oobabooga folder and then pip install that into that directory.

**Reply by dastaten** [reply]
I'm pretty sure that's what I'm doing, but it says it's already installed in c:\users\dorian\appdata\local\programs\python\python310\lib\site-packages. I tried uninstalling and reinstalling in the oobabooga directory, and that didn't work either.

1. I ran 'pip uninstall sentence_transformers'
2. sentence_transformers successfully uninstalled
3. I ran 'pip install sentence_transformers' in E:\oobabooga_windows
4. It said the requirements were already satisfied

![image](https://user-images.githubusercontent.com/3282357/236613850-dac180d3-e6ab-4f1b-8b6c-b16ccb2d8019.png)

**Reply by PalaSerialX** [reply]
yeah the problem is that the library is not finding its path to the module so there is 2 things adding it to the path or installing it directly where the long term memory script.py is, its annoying but yeah once you get it on that path so it can connect with it.

**Reply by dastaten** [reply]
How do I do that?

**Reply by PalaSerialX** [reply]
so try this first same as you were doing before going to the oobabooga but instead of just targeting that folder you will go deeper so you need to find the long term memory so it will be inside extensions folder, and then find longer term memory folder, and then you'll see inside script.py you want to target that folder and then pip install into there.

**Reply by dastaten** [reply]
1. I ran 'pip uninstall sentence_transformers'
2. sentence_transformers successfully uninstalled
3. I tried uninstalling sentence_transformers again, just to make sure it was gone, and it said it couldn't find it
4. I drilled down to E:\oobabooga_windows\text-generation-webui\extensions\long_term_memory and ran 'pip install sentence_transformers' there
5. It installed, but it said 'Requirement already satisfied' for everything and installed in c:\users\dorian\appdata\local\programs\python\python310\lib\site-packages

Is it because of this line?:

Using cached sentence_transformers-2.2.2-py3-none-any.whl

Is it just using cache, and so reinstalling in the old location?

Can I just edit a line in one of the Long-Term Memory files to adjust the path to the transformers library?

![image](https://user-images.githubusercontent.com/3282357/236615286-fc97dc3d-b104-4241-9bdd-294ba0de7fd1.png)

**Reply by PalaSerialX** [reply]
So basically, its install the library into the "general" c:\users\dorian\appdata\local\programs\python\python310\lib\site-packages where all your pip installed packages are stored.. but the extension module is not on that path, so it doesn't know that this files are in that location. So it can access them, and every time you install it, it goes into the same directory hence it says its already there. 

First, navigate to the project directory:



cd E:\oobabooga_windows\text-generation-webui\extensions\long_term_memory
Create a virtual environment using Python's built-in venv module:


python -m venv venv
This will create a folder named venv in the current directory, which contains a separate Python environment.

Activate the virtual environment:

For Windows:

.\venv\Scripts\activate
For macOS and Linux:


source venv/bin/activate
Once the virtual environment is activated, your command prompt should change, indicating that you're using the virtual environment.

Now, install the sentence_transformers package:



pip install --no-cache-dir sentence_transformers
The --no-cache-dir flag ensures that pip won't use cached packages and will download the package again.

After completing these steps, the sentence_transformers package should be installed in the venv folder inside your project directory. You don't need to adjust any paths in your project files, as the virtual environment will take care of that.

To deactivate the virtual environment once you're done, simply run:

Copy code
deactivate

**Reply by dastaten** [reply]
I did all of that, and I now have a 'venv' folder inside the 'long_term_memory' folder, and it contains 'Include', 'Lib', 'Scripts', 'Share', and 'pyvenv.cfg'. However, I'm getting what looks like the same error as before when trying to load the model.

![image](https://user-images.githubusercontent.com/3282357/236616387-092e0c7a-3324-42a7-9260-d0ceff1f339f.png)

**Comment by PalaSerialX** [comment]
okay now your entering bug territory because, I had the same issue except, I was trying to add new module into the API script.py and when running the script via command line, it gave me the same issue where it cannot find the modules and even when I tried to add code to the existing modules that it already was pulling functions from, it would not pull my functions from them, it would tell me it did not find them. I would put it in as bug report on oobabooga, unfortunately I don't have a fix for that bug through the .bat launch. I don't know why its doing that, and I tried many workarounds, loading it through Pycharm IDE is the only way to get them to work or see the modules but it can presents new problems along the way.

**Reply by dastaten** [reply]
Understood. Thanks for all your help.

**Reply by miles-du** [reply]
Hi, as a noob - I was wondering if the memory extension works with the latest updated Ooba, or do we have to use the version of Ooba that these Developers link to in their repo page, (i.e. the one they based this fork on in the first place)?
